module: remove require('.') with NODE_PATH compat

silverwind opened the pull request:
This PR replaces #1452. It's updated and targets `master` now. Adding 6.0.0 milestone as discussed.

silverwind commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/514/

silverwind commented:
Going to postpone this removal once more. This incompatibilty and the subsequent hack were added during io.js. the exposure time of this deprecation message would be quite short for users of 4.0.0. Likely target is 7.0.0.

jasnell commented:
@silverwind ... is this still something you want to pursue?

silverwind commented:
Yes, I'm just waiting for the v6 release so I can land this on master.

benjamingr commented:
Mind elaborating on this change? Also, can we get a citgm run?

silverwind commented:
Rebased.

@benjamingr:

When we introduced `require('.')` it had broken a undocumented use case (https://github.com/nodejs/node/issues/1356) where `NODE_PATH` was set to a module's directory (as opposed to the recommended way of setting it to a directory containing module directories).

I introduced a 'hack' to restore the functionality with a deprecation message in place (https://github.com/nodejs/node/pull/1363), which is now removed again. When this goes into the 7.0.0 release, the deprecation will have been in place for 1.5 years.

silverwind commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/2465/

ChALkeR commented:
Perhaps we should review the stability levels in the documentation first?
Filed #6528 for that.

silverwind commented:
Test is broken. Looks like something in https://github.com/nodejs/node/commit/ae18bbef48d87d9c641df85369f62cfd5ed8c250 made my "maintain backwards compat" section neccessary to resolve `'.'` correctly, investigating.

silverwind commented:
Condition fixed, new CI: https://ci.nodejs.org/job/node-test-pull-request/2468/

silverwind commented:
CI is green except an unrelated java failure. I don't think a CITGM would bring any more insights here, because modules usually don't set `NODE_PATH`.

silverwind commented:
@TheAlphaNerd can you run this through a CITGM, just to be sure?

TheAlphaNerd commented:
citgm: https://ci.nodejs.org/view/Node.js-citgm/job/thealphanerd-smoker/273/

ChALkeR commented:
Note that this change currently directly contradicts with the documentation, which says that only security, performance, or bug fixes will be accepted for `module`.

Either the documentation should be fixed first, or this change rejected.

silverwind commented on lib/module.js:
```diff
@@ -285,12 +276,11 @@ Module._resolveLookupPaths = function(request, parent) {
+  if (request.charCodeAt(0) !== 46/*.*/ || (
+      request.charCodeAt(1) !== 46/*.*/ &&
+      request.charCodeAt(1) !== 47/*/*/ &&
+      request.length > 1)) {
```
@mscdex does this change look right to you? The condition was `start !== '.' && start !== './' && start !== '..'` before you refactored it to use `charCodeAt`. Tests are passing, but I'd like to get a second look on that one.

mscdex replied:
You'll want to add request length checking *before* using `charCodeAt(n)` to make sure you don't cause a deopt due to an out-of-bounds access.

silverwind commented:
While https://github.com/nodejs/node/issues/6528 is still in limbo, I'd like to get this hack removed rather sooner than later.

@nodejs/collaborators any concerns regarding the changes? (See https://github.com/nodejs/node/pull/3384#issuecomment-216338572 for a summary)

mscdex commented on test/parallel/test-require-dot.js:
```diff
@@ -1,17 +1,9 @@
-
-var c = require('.');
-
-assert.equal(c.value, 42, 'require(".") should honor NODE_PATH');
```
Instead of removing this test should we merely assert the opposite?

silverwind replied:
Good idea!

silverwind commented:
Totally forgot about this one. I'll move it to the next milestone. Has kind of an benefit in that it won't be introduced in a LTS cycle.

silverwind commented:
Yes, rebased and simplified the relative path check. Turns out the previous length check was obsolete so I was able to remove it. Also altered the test to test for the opposite as to @mscdex's suggestion.

silverwind commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/5930/
CITGM: https://ci.nodejs.org/job/citgm-smoker/537/

jasnell commented:
@nodejs/ctc ... thoughts on this?

silverwind commented:
Still looking for reviews here. I'd say its a very low impact removal of a long deprecated hack.

Maybe @Fishrock123 @rvagg could have a look as they reviewed when the hack landed in https://github.com/nodejs/node/commit/3ad82c335deff29a1db5a391f893dc6d2ff1e63a?

evanlucas commented:
IMO, @ChALkeR's concern still stands:

> Note that this change currently directly contradicts with the documentation, which says that only security, performance, or bug fixes will be accepted for module.

> Either the documentation should be fixed first, or this change rejected.

silverwind commented:
I'm not even sure how the wording on such a doc update should be. I'd classify this PR as "Removal of hacks for features that were never indented/documented to work the way they did".

silverwind commented:
`modules` has been [downgraded to stable](https://github.com/nodejs/node/commit/9bc042cb7431e507b5c8b77689a9264101650d42), which should enable this PR to be landed.

jasnell commented:
@nodejs/ctc ... I have put this on for `ctc-review`. Is this something we want to do for 8.0.0 or no? Is it something we want to do at all?

BridgeAR commented:
This was taken from the [ctc-review](https://github.com/nodejs/CTC/issues/130#issuecomment-302874305) and I agree with @silverwind that after all this time passed since the deprecation there should be little risk to land this.

Trott commented:
Now that the stability index has been updated, this change would no longer violate the stability index. We should probably make a decision on this one way or another.

silverwind commented:
Will rebase soon. I see two options to forward this issue:

1) Remove the deprecation and accept that someone should be able to do `NODE_PATH=/path/to/module`, do `require('.')` and get the module instead of one in `$PWD` (which is imho unexpected).

2) Remove both the deprecation and the hack. Might also bring a tiny bit of performance when loading modules.

Trott commented:
@nodejs/ctc: Thoughts on this one? Needs a CITGM run, but assuming that comes up with no surprises, is this something we want to do? (And if so, which of the two approaches @silverwind proposes in the preceding comment?) And if this is something we don't ever want to do, we should close this. :-D

targos commented:
I  am -1 for option 1, +1 for option 2.

jasnell commented:
+1 for option 2

jasnell commented:
ping @nodejs/ctc ... please review

BridgeAR commented:
+1 for option 2

silverwind commented:
I don't assume that a CITGM will catch anything here. The compat is just for users that set `NODE_PATH` manually, which I assume no sane module would do by itself. Oh, and those users now have been seeing the deprecation for almost 2.5 years. I'll start a CI run now.

silverwind commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9878/

(For real now, couldn't start the last time because the CI was down)

silverwind commented:
@jasnell done.

I'll land this tomorrow if there are no objections.

jdalton commented:
🎉  which Node version is this expected to land in?

jasnell commented:
It's a major so it would land in 9.0.0

silverwind commented:
Finally landed in https://github.com/nodejs/node/commit/a517466aa7dcb7afe4864ab12d0f97e10a8d4ee0. It's been a long ride.

MylesBorins commented:
@nodejs/tsc it appears that this landed without running citgm and it is breaking a fairly insignificant portion of our ecosystem.

It breaks a common use case of the module [d](https://github.com/medikoo/d) specifically:

```js
require('d/lazy')
```

A revert is incoming. If we had caught this earlier we may have been able to come up with a fix but with d receiving over 500k downloads a day I don't see how we can ship this tomorrow

MylesBorins commented:
Some more details after a bit more research

this commit specifically broke the case of requiring individual files in modules that only have a single character as a name. This will thankfully be limited to 26 cases!

The specific line that broke things is https://github.com/nodejs/node/commit/a517466aa7dcb7afe4864ab12d0f97e10a8d4ee0#diff-d1234a869b3d648ebfcdce5a76747d71L338

In light of this specific edge case we may want to re-examine the deprecation all together
